src/Controls/src/Core/Platform/ModalNavigationManager/ModalNavigationManager.cs

```diff
@@ -9,52 +10,261 @@ namespace Microsoft.Maui.Controls.Platform
+			// If the window handler is changing the activity is being recreated
+			// the window activated/resumed event will take care of syncing the platform modals
+			if (e.OldHandler is not null)
+			{
+				_platformModalPages.Clear();
+			}
 		}
 
+		public Task<Page?> PopModalAsync()
+		{
+			return PopModalAsync(true);
+		}
 
 		public Task PushModalAsync(Page modal)
 		{
 			return PushModalAsync(modal, true);
 		}
 
-		internal void SettingNewPage()
+		bool syncing = false;
+
+		bool IsModalReady
+		{
+			get
+			{
+				return
+#if ANDROID
+					_window.IsActivated &&
+#endif
+					_window?.Page?.Handler != null && _window.Handler != null;
+			}
+		}
+
+		void SyncPlatformModalStack([CallerMemberName] string? callerName = null)
+		{
+			var logger = _window.FindMauiContext(true)?.Services?.CreateLogger<ModalNavigationManager>();
+			SyncPlatformModalStackAsync().FireAndForget(logger, callerName);
+		}
+
+		// This code only processes a single sync action per call.
+		// It recursively calls itself until no more sync actions are left to perform.
+		//
+		// A lot can change during the process of pushing/popping a page
+		// i.e. Users might change the root page during an appearing event.
+		// So, instead of just bull dozing through the whole sync we perform one
+		// sync step then recalculate the state of affairs and then perform another
+		// until no more sync operations are left.
+		// Typically it's always a good idea to re-evaluate after any async operation has completed
+		async Task SyncPlatformModalStackAsync()
+		{
+			if (!IsModalReady || syncing)
+				return;
+
+			bool changed = false;
+
+			try
+			{
+				syncing = true;
+
+				await WindowReadyForModal();
+
+				int popTo;
+
+				for (popTo = 0; popTo < _platformModalPages.Count && popTo < _modalPages.Count; popTo++)
+				{
+					if (_platformModalPages[popTo] != _modalPages[popTo])
+					{
+						break;
+					}
+				}
+
+				// This means the modal stacks are already synced so we don't have to do anything
+				if (_platformModalPages.Count == _modalPages.Count && popTo == _platformModalPages.Count)
+					return;
+
+				// This ensures that appearing has fired on the final page that will be visible after 
+				// the sync has finished
+				CurrentPage?.SendAppearing();
+
+				// Pop platform modal pages until we get to the point where the xplat expectation
+				// matches the platform modals
+				if (_platformModalPages.Count > popTo && IsModalReady)
+				{
+					var page = await PopModalPlatformAsync(false);
+					page.Parent = null;
+					changed = true;
+				}
+
+				if (!changed)
+				{
+					//push any modals that need to be synced
+					var i = _platformModalPages.Count;
+					if (i < _modalPages.Count && IsModalReady)
+					{
+						await PushModalPlatformAsync(_modalPages[i], false);
+						changed = true;
+					}
+				}
+			}
+			finally
+			{
+				syncing = false;
+			}
+
+			if (changed)
+			{
+				await SyncPlatformModalStackAsync();
+			}
```

Comment:
If there's an exception inside the try/clause the entire app will probably crash, plus we probably don't want to recursively call this method again if it crashed.  It needs to just exit from trying to sync and pass the exception up to the wherever this sync started from. 

Also, I have it outside the try catch because I want this finally to run and set syncing back to false, before I recurse back into the sync method.

![image](https://user-images.githubusercontent.com/5375137/217659055-48a59414-8c28-45ea-a953-42a0cca2a06e.png)

Reply:
We got a call, understood, we want the syncing var to always be set to false EVEN when we do have a return statement inside the try {} it has nothing to do with a possible exception. Adding my +1